x/liquidstakeibc/keeper/ica.go

```diff
@@ -0,0 +1,100 @@
+package keeper
+
+import (
+	"fmt"
+
+	errorsmod "cosmossdk.io/errors"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/cosmos/ibc-go/v6/modules/apps/27-interchain-accounts/controller/types"
+	icatypes "github.com/cosmos/ibc-go/v6/modules/apps/27-interchain-accounts/types"
+	"github.com/gogo/protobuf/proto"
+
+	liquidstakeibctypes "github.com/persistenceOne/pstake-native/v2/x/liquidstakeibc/types"
+)
+
+const (
+	PacketSrcChannelEvent string = "packet_src_channel"
+)
+
+func (k *Keeper) GenerateAndExecuteICATx(
+	ctx sdk.Context,
+	connectionID string,
+	ownerID string,
+	messages []proto.Message,
+) (string, error) {
+
+	msgData, err := icatypes.SerializeCosmosTx(k.cdc, messages)
+	if err != nil {
+		k.Logger(ctx).Error(fmt.Sprintf("could not serialize tx data: %v", err))
+		return "", err
+	}
+
+	icaPacketData := icatypes.InterchainAccountPacketData{
+		Type: icatypes.EXECUTE_TX,
+		Data: msgData,
+	}
+
+	msgSendTx := &types.MsgSendTx{
+		Owner:           ownerID,
+		ConnectionId:    connectionID,
+		PacketData:      icaPacketData,
+		RelativeTimeout: uint64(liquidstakeibctypes.ICATimeoutTimestamp.Nanoseconds()),
+	}
+
+	handler := k.msgRouter.Handler(msgSendTx)
+	res, err := handler(ctx, msgSendTx)
+	if err != nil {
+		k.Logger(ctx).Error(fmt.Sprintf("sending ica tx with msg: %s failed with err: %v", msgData, err))
+		return "", errorsmod.Wrapf(liquidstakeibctypes.ErrICATxFailure, "failed to send ica msg with err: %v", err)
+	}
+	ctx.EventManager().EmitEvents(res.GetEvents())
+
+	// retrieve the channel id from the result events
+	channelID := ""
+	for _, event := range res.Events {
```

Comment:
can we not depend on events? they are not part of consensus so far , so they can be changed anytime

Reply:
I did it this way because apparently couldn't find any other way of obtaining the channel id the tx was sent over...